akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -422,6 +425,10 @@ class ClusterSingletonManager(
     removed = removed filter { case (_, deadline) ⇒ deadline.hasTimeLeft }
   }
 
+  // for CoordinatedShutdown
+  val memberExitingProgress = Promise[Done]()
+  CoordinatedShutdown(context.system).addTask("cluster-pre-exiting")(() ⇒ memberExitingProgress.future)
```

Comment:
changed my mind, I'll introduce configurable timeouts for the phases and a recover flag to indicate if it should continue in case of exception